src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -232,11 +241,150 @@ export function getMostRecentBundles(
-//   // 1. First try to fetch flow information from existing bundle state saved in client memory.
-//   // 2. If not available, then we need to load additional data from the RPC.
-//   throw new Error("Unimplemented");
-// }
+export async function getModifiedFlow(
+  chainId: number,
+  flow: UbaFlow,
+  hubPoolClient: HubPoolClient,
+  spokePoolClients: SpokePoolClients
+): Promise<ModifiedUBAFlow> {
+  let tokenSymbol: string | undefined;
+  if (isUbaInflow(flow)) {
+    if (chainId !== flow.originChainId) {
+      throw new Error(`ChainId ${chainId} should be deposit.originChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(flow.originChainId, flow.originToken)?.symbol;
+  } else if (outflowIsFill(flow as UbaOutflow)) {
+    if (chainId !== flow.destinationChainId) {
+      throw new Error(`ChainId ${chainId} should be fill.destinationChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(flow.destinationChainId, (flow as FillWithBlock).destinationToken)?.symbol;
+  } else if (chainId !== (flow as RefundRequestWithBlock).repaymentChainId) {
+    if (chainId !== flow.repaymentChainId) {
+      throw new Error(`ChainId ${chainId} should be refund.repaymentChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(
+      flow.repaymentChainId,
+      (flow as RefundRequestWithBlock).refundToken
+    )?.symbol;
+  }
+  if (!tokenSymbol) {
+    throw new Error(`Could not find token symbol for chain ${chainId} and flow ${JSON.stringify(flow)}`);
+  }
+
+  // Load bundle ranges before flow and until after it:
+  const bundleRanges = Object.fromEntries(
+    Object.keys(spokePoolClients).map((_chainId) => {
+      const bundles = getMostRecentBundleBlockRanges(
+        Number(_chainId),
+        // 3 seems like a reasonable number of bundles to lookback to guarantee we can validate all events in the
+        // most recent one.
+        3,
+        flow.quoteBlockNumber,
+        hubPoolClient,
+        spokePoolClients
+      );
+      // Make bundle.end cover from the block range until the flow.block so we know the running balance right before
+      // the flow
+      bundles[bundles.length - 1].end = flow.blockNumber;
+      return [_chainId, bundles];
+    })
+  );
+
+  // Instantiate new spoke pool clients that will look back at older data:
+  const newSpokePoolClients = Object.fromEntries(
+    await Promise.all(
+      Object.keys(spokePoolClients).map(async (_chainId) => {
+        const spokeChain = Number(_chainId);
+        // Span spoke pool client event searches from oldest bundle's start to newest bundle's end:
+        const spokePoolClientSearchSettings = {
+          fromBlock: bundleRanges[spokeChain][0].start,
+          toBlock: bundleRanges[spokeChain][bundleRanges[spokeChain].length - 1].end,
+          maxBlockLookBack: spokePoolClients[spokeChain].eventSearchConfig.maxBlockLookBack,
+        };
+        const newSpokeClient = new SpokePoolClient(
+          new Logger(),
```

Comment:
Does this work? I thought `winston`'s logger is abstract and can't be directly instantiated.

Reply:
replaced with jsut passing `spokePoolCLients[chainId].logger` :)